library/common/engine.cc

```diff
@@ -93,14 +94,10 @@ Engine::~Engine() {
   main_thread_.join();
 }
 
-void Engine::recordCounter(std::string elements, uint64_t count) {
-  if (server_) {
+void Engine::recordCounter(const std::string& elements, uint64_t count) {
```

Comment:
now that I see this in its current form, it's worth a quick check whether 'elements' value gets copied into the lambda provided to post, or whether it holds a ref to an object that might potentially not live long enough for the dispatcher to execute the post.

It might not matter in your unit test, where the lifetime of the string you pass to recordCounter is long enough.

In any case I think the interface to this function should be a const ref, but you may need to copy it explicitly into storage saved in the lambda. Can you check?

Reply:
Thanks @jmarantz, good callout.

Reply:
Yeah good point! 🤦 the string could definitely be gone as the copy is just in the stack of the main interface function call. What's the least egrigious way to minimize copies:
- pass down the const char* all the way until we capture a copy in the lambda,
- create a unique ptr of a string at the main interface level that we pass down,
- something else?

Reply:
My recommendation:
 * leave the interface as is (or better yet, witch to a string_view to avoid copying while calling recordValue)
 * make a local copy in the function body (`std::string name = elements;`)
 * capture that copy by value in the lambda passed to `post()`.

Reply:
My apologies, what you have is fine. I tested it with:
```
TEST(Foo, Bar) {
  std::string foo = "foo";
  std::string& fooref = foo;
  auto test = [fooref]() { EXPECT_EQ("foo", fooref); };
  test();
  foo = "bar";
  test();
}
```

That test passes, which means that the ref's value is captured by the lambda.